lib/src/__tests__/Validation.test.tsx

```diff
@@ -0,0 +1,151 @@
+import * as React from 'react';
+import { isWeekend } from 'date-fns';
+import { TextField } from '@material-ui/core';
+import { MaterialUiPickersDate } from '../typings/date';
+import { DesktopDatePicker } from '../DatePicker/DatePicker';
+import { mountPickerWithState, utilsToUse } from './test-utils';
+import { DesktopDateRangePicker } from '../DateRangePicker/DateRangePicker';
+import { TimePickerProps, DesktopTimePicker } from '../TimePicker/TimePicker';
+
+const disableWeekends = (date: MaterialUiPickersDate) => {
+  return isWeekend(utilsToUse.toJsDate(date));
+};
+
+describe('DatePicker validation', () => {
+  test.each`
+    props                                     | input            | expectedError
+    ${{}}                                     | ${'invalidText'} | ${'invalidDate'}
+    ${{ disablePast: true }}                  | ${'01/01/1900'}  | ${'disablePast'}
+    ${{ disableFuture: true }}                | ${'01/01/2050'}  | ${'disableFuture'}
+    ${{ minDate: new Date('01/01/2000') }}    | ${'01/01/1990'}  | ${'minDate'}
+    ${{ maxDate: new Date('01/01/2000') }}    | ${'01/01/2010'}  | ${'maxDate'}
+    ${{ shouldDisableDate: disableWeekends }} | ${'04/25/2020'}  | ${'shouldDisableDate'}
+  `('Should dispatch onError $expectedError', ({ props, input, expectedError }) => {
+    if (process.env.UTILS === 'luxon') {
+      return;
+    }
+
+    const onErrorMock = jest.fn();
+    const component = mountPickerWithState(utilsToUse.date(), stateProps => (
+      <DesktopDatePicker {...stateProps} {...props} onError={onErrorMock} />
+    ));
+
+    component.find('input').simulate('change', {
+      target: {
+        value: input,
+      },
+    });
+
+    expect(onErrorMock).toBeCalledWith(expectedError, expect.anything());
+  });
+
+  test('It should properly annulate the error', () => {
+    if (process.env.UTILS === 'luxon') {
+      return;
+    }
+
+    const onErrorMock = jest.fn();
+    const component = mountPickerWithState(utilsToUse.date(), stateProps => (
+      <DesktopDatePicker {...stateProps} disablePast onError={onErrorMock} />
+    ));
+
+    component.find('input').simulate('change', {
```

Comment:
I am not using testing lib in pickers yet

Reply:
What about introducing it then? So we stop investing time into a direction we know we want to go back from?

Reply:
Probably you are right. I thought that will be able to move 💯 in cypress, but it will not be possible.

Reply:
I had a look at the current state of the tests on the date picker components. It seems that, we have:

- 63 with enzyme mount/shallow (much less frequent) (`*.test.tsx` + ` it('`)
- 27 with Cypress (`*.spec.ts` + ` it('`)
- 16 with date management (`*.test.ts` + ` it('`)

*What's the difference between the usage of "test(" and "it("? It seems to be a simple alias https://jestjs.io/docs/en/api#testname-fn-timeout, what about we update the codebase to a single approach to ease the static analysis of the codebase? The usage of it is more frequent, so in favor of using it everywhere.* 

For comparison, the Autocomplete component has 82 tests, and the core components have 1956 tests. So it seems that migrating the date picker components in any direction will be a much smaller effort (1/10) than the one we have currently in progress from enzyme to testing-library for the core (there are 600 tests+ still to be migrated).

Reply:
@eps1lon Assuming the sources of the date picker will be hosted on the mono-repository. Do you have any preference on the strategy we use for testing it?

Reply:
I am not sure how you calculate test count. But we have 77 cypress tests (count by cypress dashboard comment)

Reply:
> I am not sure how you calculate test count. But we have 77 cypress tests (count by cypress dashboard comment)

@dmtrKovalenko Interesting, it's the first time I see the Cypress dashboard 👌. I have ignored (1.) `App.spec.ts` and (2.) `VisualRegression.spec.ts` in the previous count. I believe we could remove 1. per the structure of the docs in the mono-repository and need to replace 2. with Argos-CI fixtures screenshot tests. 

Actually, it's interesting to look at the Cypress dashboard. Any idea why it takes 26s to run 27 tests? If we were to write all our tests with Cypress, would it mean 1956 / 27 * 26 = 1883s (31min) of test duration per browser?
In the mono-repository test stack, we run them in about 30s per browser (we run them in 4 browsers). I suspect that fixing #1674 can speed up the tests by a factor x3-5. It seems that we are still talking about **one order of magnitude** of slowdown if we make Cypress a systematic choice.

Reply:
We don't want to move all component test to cypress. That is for 99% of our tests overkill (we're talking about 1s per test instead of 50ms). Especially considering they polyfill a lot of browser behavior (we want to control). 

There are probably some tests that might be more suited for e2e tests but I'd rather use a lower level library for that (like puppeteer or playwright). Cypress really shines once you have complete user stories e.g. login or data fetching.

Reply:
In this case, I would be in favor of migrating **all** the tests to testing-library, until we hit a roadblock, something that is compelling enough to introduce a new tool in the stack, that the engineers working on Material-UI need to learn and maintain.

We had one case like this in the past, to test the UMD release. We had multiple regressions, we couldn't use Karma, we went with Puppeteer (the test is still present).

As far as I can envision the limits of testing-library + karma, we have:

- No way to test the tab behavior. I have tested this diff, using https://github.com/WordPress/gutenberg/pull/5618 as inspiration, it seems to work with Puppeteer.

```diff
diff --git a/packages/material-ui/test/umd/run.js b/packages/material-ui/test/umd/run.js
index 367466663..8c34ecb2b 100644
--- a/packages/material-ui/test/umd/run.js
+++ b/packages/material-ui/test/umd/run.js
@@ -71,6 +71,8 @@ function App() {

   return (
     <React.Fragment>
+      <input type="text" id="foo" />
+      <input type="text" id="bar" />
       <Button onClick={() => setOpen(true)}>Super Secret Password</Button>
       <Dialog open={open}>
         1-2-3-4-5
@@ -141,8 +143,12 @@ async function run() {

     const page = await startBrowser();
     await page.goto(`http://${host}:${port}`);
+    await page.click( '#foo' );
+    await page.keyboard.press( 'Tab' );
+    const activeElement = await page.evaluate(() => {
+      return document.activeElement.id;
+    });
+    console.log('activeElement', activeElement); // output bar
    await expect(page).toClick('button', { text: 'Super Secret Password' });
    await expect(page).toMatch('1-2-3-4-5');
   } catch (err) {
     log.fatal({
       name: 'test',
```

1. No way to test drag and drop, it's going to be important for the DataGrid and TreeView. For the ClockView of the date picker, it's might also be a concern but I think a less important one. There are no underlying elements that could catch the events and mess around.
1. No reliable way to test the performance of the components (do we?). It's going to be important for the DataGrid. We might also need to track this metric too for #1674.
1. No built-in parallelism. We can probably still write x2 more tests without it harming our DX. After that, we might have an issue with the time it takes for the tests to run. We could split the tests into two chunks, one half going to BrowserStack, the other to SauceLabs.
1. else? did I forget something? The debugging experience isn't great but it doesn't seem to have been something we have struggled a lot with (so far).

Reply:
> No reliable way to test the performance of the components (do we?)

I experimented a bit with React.Profiler in our test suite which means that we wouldn't need to write tests and benchmark but reuse the testcases for benchmarking. That's a viable option if I remember correctly. Just needs some infra so that we get a good sample size and don't run it on any PR.

> No way to test drag and drop, it's going to be important for the DataGrid and TreeView.

That is something I wouldn't let automated testing touch anyway. Fixtures and manual testing would be way more reliable and also ensures that Drag'n'Drop is actually usable. I think this should be tested manually on different devices.

> We can probably still write x2 more tests without it harming our DX.

Please consider that your workflow is the issue. You absolutely do not need to run the full test suite in watchmode when working on the codebase for virtually every use case. `--grep` and `.only` give you all the tools you need.

> No way to test the tab behavior.

This is something I never understood. The only meaningful behavior you're missing out is :focus-visible and that is only visual anyway. For all the rest `.focus()` should be enough, no? If you're interested in tab-order then you need to test this manually with an actual browsers/assistive technology.

Reply:
@eps1lon 

1. Great. I have no knowledge of how this should be done correctly. My understanding of the problem is limited to: reproducibility is paramount. Without a stable, noise-free measure, we can't do anything.
2. Agree, manual testing should probably be the main tool used when iterating on the logic. Maybe one automated test, as a rough safe belt could be valuable.
3. My concern is with the CI, the **delay** of the feedback loop on pull requests.
4. We consider the TrapFocus component, do we have enough tools to test it correctly?